### PR TITLE
"Add ID column" Button for Migration Editor

### PR DESCRIPTION
**Pull request by DanHarrin**
I always see myself repeating the same few steps on all of my models to add an ID column, so I added a button to do it for me - just to save that little bit of time.

Screenshot

### CODE REVIEW

File: assets/js/builder.index.entity.databasetable.js [hunk 1]
```diff
@@ -86,6 +86,15 @@
         this.unmodifyTab($masterTabPane)
     }
 
+    DatabaseTable.prototype.cmdAddId = function(ev) {
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin Would you be able to rename this function to `cmdAddIdColumn`?

File: assets/js/builder.index.entity.databasetable.js [hunk 2]
```diff
@@ -159,8 +168,12 @@
         var $masterTabPane = this.getMasterTabsActivePane(),
             $form = $masterTabPane.find('form'),
             $toolbar = $masterTabPane.find('div[data-control=table] div.toolbar'),
-            $button = $('<a class="btn oc-icon-clock-o builder-custom-table-button" data-builder-command="databaseTable:cmdAddTimestamps"></a>')
+            $button = $('<a class="btn oc-icon-clock-o builder-custom-table-button" data-builder-command="databaseTable:cmdAddId"></a>')
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin The re-use of `$button` is a little bit of a code smell - would you be able to define this new button and the original "Add Timestamps" button with different variables? That way you can include them both in the `var` definition list.

File: assets/js/builder.index.entity.databasetable.js [hunk 3]
```diff
@@ -257,6 +270,42 @@
         return result
     }
 
+    DatabaseTable.prototype.addIdColumn = function($target) {
+        var existingColumns = this.getColumnNames($target),
+            added = false
+
+        if ($.inArray('id', existingColumns) == -1) {
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin Any particular reason for jQuery's `inArray` over the native `indexOf`?

File: assets/js/builder.index.entity.databasetable.js [hunk 4]
```diff
@@ -257,6 +270,42 @@
...
+                currentData = this.getTableData($target),
+                rowData = {
+                    name: 'id',
+                    type: 'integer',
+                    unsigned: true,
+                    auto_increment: true,
+                    primary_key: true,
+                }
+
+            if (currentData.length-1) {
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin Spaces needed:
```suggestion
            if (currentData.length - 1) {
```

File: assets/js/builder.index.entity.databasetable.js [hunk 5]
```diff
@@ -257,6 +270,42 @@
...
+                tableObj.addRecord('bottom', true)
+            }
+
+            tableObj.setRowValues(currentData.length-1, rowData)
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin  Spaces needed:
```suggestion
            tableObj.setRowValues(currentData.length - 1, rowData)
```

File: lang/en/lang.php [hunk 6]
```diff
@@ -57,8 +57,10 @@
         'tab_new_table' => 'New table',
         'btn_add_column' => 'Add column',
         'btn_delete_column' => 'Delete column',
+        'btn_add_id' => 'Add ID',
         'btn_add_timestamps' => 'Add timestamps',
         'btn_add_soft_deleting' => 'Add soft deleting support',
+        'id_exists' => 'id column already exists in the table.',
```
Comment (bennothommo) [inline lang/en/lang.php]: @DanHarrin Would you be able to capitalise the `ID` in the message here?

File: updates/version.yaml [hunk 7]
```diff
@@ -23,3 +23,4 @@
 1.0.23: Added support for balloon-selector field type, added Brazilian Portuguese translation, fixed some bugs
 1.0.24: Added support for tag list field type, added read only toggle for fields. Prevent plugins from using reserved PHP keywords for class names and namespaces
 1.0.25: Allow editing of migration code in the "Migration" popup when saving changes in the database editor.
+1.0.26: Added migration editor button to add an 'id' column to a table.
```
Comment (bennothommo) [inline updates/version.yaml]: @DanHarrin I will combine this update with the other updates waiting for the next version, so could you remove the `updates/version.yaml` change for us?

**Comment by DanHarrin** [comment]
All done @bennothommo

**Comment by bennothommo** [comment]
@DanHarrin I have tested this - it works fine with no columns inserted, but if you were to enter a column (say a string) first, then click the "Add ID" button, it overwrites the first column. See below:

ID Column Issue

**Comment by DanHarrin** [comment]
@bennothommo @w20k that issue should be fixed with my most recent commit.

File: assets/js/build-min.js [hunk 8]
```diff
@@ -167,10 +169,19 @@ tableObj.unfocusTable()
 var data=this.getTableData($target),result=[]
 for(var index in data){if(data[index].name!==undefined){result.push($.trim(data[index].name))}}
 return result}
+DatabaseTable.prototype.addIdColumn=function($target){var existingColumns=this.getColumnNames($target),added=false
+if(existingColumns.indexOf('id')==-1){var tableObj=this.getTableControlObject($target),currentData=this.getTableData($target),rowData={name:'id',type:'integer',unsigned:true,auto_increment:true,primary_key:true,}
+console.log(currentData);if(currentData.length-1||currentData[0].name||currentData[0].type||currentData[0].length||currentData[0].unsigned||currentData[0].nullable||currentData[0].auto_increment||currentData[0].primary_key||currentData[0].default){tableObj.addRecord('bottom',true)}
```
Comment (w20k) [inline assets/js/build-min.js]: @DanHarrin why `console.log` is in compiled files, any idea? :)

**Comment by DanHarrin** [comment]
I apologise @w20k, that was used during testing and was not recompiled. It's been fixed.

**Comment by w20k** [comment]
@bennothommo - done, but re-testing needed!

File: assets/js/builder.index.entity.databasetable.js [hunk 9]
```diff
@@ -257,6 +270,42 @@
         return result
     }
 
+    DatabaseTable.prototype.addIdColumn = function($target) {
+        var existingColumns = this.getColumnNames($target),
+            added = false
+
+        if (existingColumns.indexOf('id') == -1) {
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin This needs to be an explicit check with `===`

File: assets/js/builder.index.entity.databasetable.js [hunk 10]
```diff
@@ -265,7 +314,7 @@
         for (var index in columns) {
             var column = columns[index]
 
-            if ($.inArray(column, existingColumns) == -1) {
+            if (existingColumns.indexOf(column) == -1) {
```
Comment (bennothommo) [inline assets/js/builder.index.entity.databasetable.js]: @DanHarrin This needs to be an explicit check with `===`

**Comment by bennothommo** [comment]
@DanHarrin tested it and it is working all good now. Just a couple of minor tweaks and I'll merge it! Well done :)

**Comment by DanHarrin** [comment]
Done the explicit checks @bennothommo

**Pull request merged**